strawberryfields/program.py

```diff
@@ -765,49 +696,99 @@ def compile(self, backend='fock', **kwargs):
-                            # simplify the error message by suppressing the previous exception
-                            raise err from None
+
+                elif op_name in db.decompositions:
+                    # backend requests an op decomposition
+
+                    # TODO: allow the user to selectively turn off decomposition
+                    # by passing the kwarg `decomp=False` to more
+                    # operations (currently only ops.Gaussian allows this).
+                    #
+                    # For example, the 'gaussian' backend supports setting the state
+                    # via passing directly the (mu, cov) OR by first having the
+                    # frontend decompose into other primitive gates.
+                    # That is, ops.Gaussian is both a primitive _and_ a decomposition
+                    # for the 'gaussian' backend, and it's behaviour can be chosen
+                    # by the user.
+                    if (op_name in db.primitives) and hasattr(cmd.op, 'decomp'):
+                        # op is a backend primitive, AND backend also
+                        # supports decomposition of this primitive.
+                        if not cmd.op.decomp:
+                            # However, user has requested to bypass decomposition
+                            compiled.append(cmd)
+                            continue
+
+                    try:
+                        kwargs = db.decompositions[op_name]
+                        temp = cmd.op.decompose(cmd.reg, **kwargs)
+                        # now compile the decomposition
+                        temp = compile_sequence(temp)
+                        compiled.extend(temp)
+                    except NotImplementedError as err:
+                        # Operation does not have _decompose() method defined!
+                        # simplify the error message by suppressing the previous exception
+                        raise err from None
+
+                elif op_name in db.primitives:
+                    # backend can handle the op natively
+                    compiled.append(cmd)
+
                 else:
                     raise CircuitError('The operation {} cannot be used with the {} backend.'.format(cmd.op.__class__.__name__, backend))
+
             return compiled
 
         self.lock()
         seq = compile_sequence(self.circuit)
+
+        if db.topology is not None:
+            # check topology
+            grid = self._list_to_grid(seq)
+            DAG = self._grid_to_DAG(grid)
+
+            # relabel the DAG nodes to integers, with attributes
+            # specifying the operation name. This allows them to be
+            # compared, rather than using Command objects.
+            mapping = {i: n.op.__class__.__name__ for i, n in enumerate(DAG.nodes())}
+            circuit = nx.convert_node_labels_to_integers(DAG)
+            nx.set_node_attributes(circuit, mapping, name='name')
```

Comment:
Getting a bit confused. Does 'name' refer to the generic name of the Command (e.g., 'BSgate') or to a unique name associated with each node?

Reply:
You're correct, name refers to the operation name `BSgate`.

Reply:
Ok, so this code checks that the graph topology matches, and that at each node, the ops have the same name. Does it check that these ops have the same parameters?

Reply:
Not yet. Currently, it just checks the name.

I tried to code this in, but it sort of bleeds into the compilation problem as well, so I removed that code. Do you want me to add it back/finish it?

Reply:
So currently two graphs are considered isomorphic if they have all the same gates in the same places, but whether arguments match or not does not influence it?

Reply:
Yes, that is not currently done - this will be done in a new ticket.